designs/vsphere-advanced-vmx-settings.md

```diff
@@ -0,0 +1,153 @@
+
+vSphere allows users to configure [advanced VMX settings](https://docs.vmware.com/en/VMware-vSphere/7.0/com.vmware.vsphere.vcenter.configuration.doc/GUID-62184858-32E6-4DAC-A700-2C80667C3558.html) for individual VMs. However, EKS Anywhere on vSphere currently doesn't provide a way for users to modify these settings.
+
+One of the use-cases for exposing this is to allow users to [disable swap space](https://github.com/aws/eks-anywhere/issues/4237) on VMs by adding this VMX config `sched.swap.vmxSwapEnabled=FALSE`.
+
+### Goals and Objectives
+
+As an EKS Anywhere user, I want to:
+
+* configure VMX settings through the EKS Anywhere clusterconfig.
+
+### Statement of Scope
+
+#### In Scope
+
+* Allow users to enable/disable VMX settings on their cluster VMs through EKS Anywhere.
+
+#### Future Scope
+
+* Iteratively add more VMX configuration options, depending upon [the approach](#option-1-opinionated-approach-expose-specific-configs).
+
+## Overview of Solution
+
+CAPV already supports configuring custom VMX keys.
+```bash
+$ kubectl explain vspheremachinetemplate.spec.template.spec.customVMXKeys
+KIND:     VSphereMachineTemplate
+VERSION:  infrastructure.cluster.x-k8s.io/v1beta1
+
+FIELD:    customVMXKeys <map[string]string>
+
+DESCRIPTION:
+     CustomVMXKeys is a dictionary of advanced VMX options that can be set on VM
+     Defaults to empty map
+```
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: VSphereMachineTemplate
+metadata:
+  name: vsphere-quickstart-worker
+  namespace: default
+spec:
+  template:
+    spec:
+      customVMXKeys:
+        "sched.swap.vmxSwapEnabled": "FALSE"
+...
+```
+
+The plan is for EKS-A to expose the same values in the EKS-A cluster.yaml.
+
+### Solution Details
+
+We can implement this in one of two ways:
+
+1. Expose some VMX options and add more incremently as users ask for it.
+1. Expose the custom VMX options as a generic map and allow users to pass in specific config themselves.
+
+#### Option 1: Opinionated approach: Expose specific configs
+
+In this approach, EKS-A will abstract away the exact VMX options and provide a user-friendly way to configure these settings.
+
+The keys and values for the options will be pre-defined types instead of generic strings.
+
+This is what the VSphereMachineConfigSpec API will look like with this approach:
+```go
+type VSphereMachineConfigSpec struct {
+  VMXOptions CustomVMXOptions `json:"vmxOptions,omitempty"`
+  ...
+}
+
+type CustomVMXOptions struct {
+  EnableSwap bool `json:"enableSwap,omitempty"`
+  ...                                            // More options can be added as needed
+}
+
+type CustomVMXOption string
+
+const (
+  EnableSwap CustomVMXOption = "sched.swap.vmxSwapEnabled"
+)
+```
+
+And the clusterconfig:
+
+```yaml
+apiVersion: anywhere.eks.amazonaws.com/v1alpha1
+kind: VSphereMachineConfig
+metadata:
+   name: my-cluster-machines
+spec:
+  diskGiB: 25
+  datastore: "Datastore"
+  folder: "Folder"
+  vmxOptions:
```

Comment:
I would recommend we put this under a field name something like `advanced` or `experimental`, etc. This will allow communication in the code that these fields are non standard, or not validated, or etc. We can also toggle all fields under this to allow for advanced use-cases and the like.